pydantic/fields.py

```diff
@@ -297,16 +324,24 @@ def _validate_list_set(
             else:
                 result.append(r)
 
+        # if no errors found, convert validated sets and tuples
+        # to their original container types, return other types as a list
+        convert_cls: Optional[type] = None
+        if self.shape is Shape.SEQUENCE and isinstance(v, tuple):
+            convert_cls = tuple
+        elif self.shape is Shape.SET:
+            convert_cls = set
+
```

Comment:
maybe slightly cleaner to do something like

```py
if errors:
    return v, errors

if self.shape is Shape.SET:
    result = set(result)
elif self.shape is Shape.SEQUENCE:
    if isinstance(v, tuple):
        result = tuple(result)
    elif isinstance(v, set):  # <-- is this case required?
        result = set(result)
return result, None
```

Also, what about the case that v was a generator? To be technically accurate, should we use `iter()` to create a new generator to return, or is that mad?

Reply:
Oh, you are right, checking ``Shape`` after returning with errors not only cleaner, but also a tiny bit more performant.

As for ``iter()`` and ``set()`` your proposal is legit. When dealing with ``Sequence`` we do our best to try to stick to original value's type. Actually, we could even return a generator expression like ``(i for i in result)`` for ``Generator`` values. But I'm not sure if it's practical to do so. So, may be, let's stick to ``iter()`` instead? 

I've fixed the code

Reply:
Not quite sure why Codecov didn't report correctly. Locally the coverage didn't change. Actually I've added even more tests for ``Sequence``